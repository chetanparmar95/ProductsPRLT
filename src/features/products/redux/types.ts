
export const FETCH_PRODUCTS = 'FETCH_PRODUCTS';
export const FETCH_PRODUCTS_SUCCESS = 'FETCH_PRODUCTS_SUCCESS';
export const FETCH_PRODUCTS_FAILED = 'FETCH_PRODUCTS_FAILED';

export const ADD_TO_CART = 'ADD_TO_CART';
export const REMOVE_FROM_CART = 'REMOVE_FROM_CART';
export const INCREMENT_QUANTITY = 'INCREMENT_QUANTITY';
export const DECREMENT_QUANTITY = 'DECREMENT_QUANTITY';