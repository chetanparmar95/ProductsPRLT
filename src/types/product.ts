export interface Product {
    id: number;
    colour: string;
    name: string;
    price: number;
    img: string;
    quantity: number;
}