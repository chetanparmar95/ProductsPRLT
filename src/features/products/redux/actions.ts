import { Product } from "../../../types/product";
import { ADD_TO_CART, DECREMENT_QUANTITY, FETCH_PRODUCTS, FETCH_PRODUCTS_FAILED, FETCH_PRODUCTS_SUCCESS, INCREMENT_QUANTITY, REMOVE_FROM_CART } from "./types";

export const fetchProducts = () => {
    return { type: FETCH_PRODUCTS }
}

export const fetchProductsSuccess = (payload: Product[]) => {
    const products = payload.map(product => {
        return {...product, quantity: 0}
    })
    return { type: FETCH_PRODUCTS_SUCCESS, payload: products}
}

export const fetchProductsFailed = (error: any) => {
    return { type: FETCH_PRODUCTS_FAILED, payload: error }
}

export const addItemToCart = (product: Product) => {
    return { type: ADD_TO_CART, payload: product}
}

export const removeItemFromCart = (productId: number) => {
    return { type: REMOVE_FROM_CART, payload: {id: productId}}
}

export const incrementQuantity = (productId: number) => {
    return { type: INCREMENT_QUANTITY, payload: {id: productId}}
}

export const decrementQuantity = (productId: number) => {
    return { type: DECREMENT_QUANTITY, payload: {id: productId}}
}