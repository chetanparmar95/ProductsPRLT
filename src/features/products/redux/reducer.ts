import { Reducer } from "redux";
import { Product } from "../../../types/product";
import { ADD_TO_CART, DECREMENT_QUANTITY, FETCH_PRODUCTS_FAILED, FETCH_PRODUCTS_SUCCESS, INCREMENT_QUANTITY, REMOVE_FROM_CART } from "./types";

export interface ProductReducerState {
    products: Product[],
    cartItems: Product[],
    error: any
}
const initialState: ProductReducerState = {
    products: [],
    cartItems: [],
    error: null
};

const productReducer: Reducer<ProductReducerState> = (state = initialState, action: any) => {
    switch (action.type) {
        case FETCH_PRODUCTS_SUCCESS:
            const products = action.payload?.map((p: Product) => ({...p, quantity: 0}))
            return {
                ...state,
                products: products,
            };
        case FETCH_PRODUCTS_FAILED:
            return {
                ...state,
                error: action.payload,
            };
        case ADD_TO_CART:
            const exist = state.cartItems.find(item => item.id === action.payload.id)
            if(!exist) {
                const product = {...action.payload, quantity: 1}
                return {
                    ...state,
                    cartItems: [...state.cartItems, product]
                }
            }
            return { ...state }
        case REMOVE_FROM_CART:
            const filteredList = state.cartItems.filter(item => item.id !== action.payload.id)
            return {
                ...state,
                cartItems: filteredList
            }
        case INCREMENT_QUANTITY:
            const { id: incrementId } = action.payload;
            const incrementedCartItems = state.cartItems.map((item) => {
                if (item.id === incrementId) {
                    return { ...item, quantity: item.quantity + 1 };
                }
                return item;
            });
            return { ...state, cartItems: incrementedCartItems };
        case DECREMENT_QUANTITY:
            const { id: decrementId } = action.payload;
            const decrementedCartItems = state.cartItems.map((item) => {
                if (item.id === decrementId && item.quantity && item.quantity > 1) {
                    return { ...item, quantity: item.quantity - 1 };
                }
                return item;
            });
            return { ...state, cartItems: decrementedCartItems };
        default:
            return state;
    }
};

export default productReducer;