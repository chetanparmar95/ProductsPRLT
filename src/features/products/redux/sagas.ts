import { call, put, takeLatest } from 'redux-saga/effects'
import ProductServices from '../../../services/products';
import { fetchProductsFailed, fetchProductsSuccess } from './actions';
import { FETCH_PRODUCTS } from './types';

// exporting for UT purpose
export function* fetchProducts(): Generator<any> {
   try {
      const response: any = yield call(ProductServices.getProducts);
      yield put(fetchProductsSuccess(response.data));
   } catch (e: any) {
      yield put(fetchProductsFailed(e));
   }
}

function* productSaga() {
  yield takeLatest(FETCH_PRODUCTS, fetchProducts);
}

export default productSaga;