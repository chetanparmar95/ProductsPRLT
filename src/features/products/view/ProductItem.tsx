import React from 'react';
import { Image, StyleSheet, Text, TouchableOpacity, View } from "react-native"
import { Product } from '../../../types/product';
import { useDispatch } from 'react-redux';
import { addItemToCart } from '../redux/actions';

interface Props {
    item: Product
}

const ProductItem = ({item}: Props) => {
    const dispatch = useDispatch();
    return (
        <View style={styles.card}>
            <Image source={{ uri: item.img }} style={styles.image} resizeMode='contain' />
            <View style={styles.detailsContainer}>
                <Text style={styles.productName}>{item.name}</Text>
                <Text style={styles.productColor}>Color: {item.colour}</Text>
                <Text style={styles.productPrice}>Price: ${item.price}</Text>
                <TouchableOpacity style={styles.addToCartButton} onPress={() => dispatch(addItemToCart(item))}>
                    <Text style={styles.addToCartButtonText}>Add to Cart</Text>
                </TouchableOpacity>
            </View>
        </View>
    )
}

export default ProductItem

const styles = StyleSheet.create({
    card: {
        backgroundColor: '#fff',
        borderRadius: 10,
        elevation: 3,
        shadowOffset: { width: 0, height: 2 },
        shadowColor: '#333',
        shadowOpacity: 0.3,
        shadowRadius: 2,
        marginVertical: 10,
        flexDirection: 'row',
        alignItems: 'center',
        padding: 15,
      },
    image: {
        width: 100,
        height: 160, 
        borderRadius: 10,
    },
    detailsContainer: {
        flex: 1,
        marginLeft: 10,
    },
    productName: {
        fontSize: 18,
        fontWeight: 'bold',
        width: '100%', 
    },
    productColor: {
        fontSize: 14,
        color: '#888',
        marginTop: 5,
    },
    productPrice: {
        fontSize: 16,
        fontWeight: 'bold',
        color: '#FF5722', 
        marginTop: 5,
    },
    addToCartButton: {
        backgroundColor: '#FF5722',
        borderRadius: 5,
        paddingVertical: 5,
        paddingHorizontal: 10,
        marginTop: 10,
        flexDirection: 'row',
        alignItems: 'center',
    },
    addToCartButtonText: {
        color: '#fff',
        fontWeight: 'bold',
        marginLeft: 5, 
    }
});