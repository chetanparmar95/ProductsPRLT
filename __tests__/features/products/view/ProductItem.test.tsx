
import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import ProductItem from '../../../../src/features/products/view/ProductItem';
import { Product } from '../../../../src/types/product';
import { addItemToCart } from '../../../../src/features/products/redux/actions';

const mockStore = configureStore([]);

describe('ProductItem Component', () => {
  it('ProductItem renders correctly', () => {
    const item: Product = {
      id: 2,
      name: 'Product 2',
      price: 15.0,
      quantity: 0,
      colour: "Black",
      img: "https://cdn-img.prettylittlething.com/d/c/3/3/dc337260f9ecefdb99a8c8e98cd73ccb1b79cea5_cmb6804_4.jpg?imwidth=1024"
    }
    const store = mockStore({});

    const { toJSON } = render(
      <Provider store={store}>
        <ProductItem item={item} />
      </Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should dispatch an action when add to cart button is pressed', () => {
    const item: Product = {
      id: 1,
      name: 'Product 1',
      price: 10.0,
      colour: 'black',
      img: 'https://example.com/product1.jpg',
      quantity: 2,
    };

    const store = mockStore({});

    const { getByText } = render(
      <Provider store={store}>
        <ProductItem item={item} />
      </Provider>
    );

    // Find and click the decrement button
    const addButton = getByText('Add to Cart');
    fireEvent.press(addButton);

    // Check if the expected action was dispatched
    const actions = store.getActions();
    expect(actions).toContainEqual(addItemToCart(item));
  });
})
