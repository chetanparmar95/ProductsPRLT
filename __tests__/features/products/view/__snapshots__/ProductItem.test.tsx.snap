// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem Component ProductItem renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderRadius": 10,
      "elevation": 3,
      "flexDirection": "row",
      "marginVertical": 10,
      "padding": 15,
      "shadowColor": "#333",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 2,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "https://cdn-img.prettylittlething.com/d/c/3/3/dc337260f9ecefdb99a8c8e98cd73ccb1b79cea5_cmb6804_4.jpg?imwidth=1024",
      }
    }
    style={
      {
        "borderRadius": 10,
        "height": 160,
        "width": 100,
      }
    }
  />
  <View
    style={
      {
        "flex": 1,
        "marginLeft": 10,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
          "width": "100%",
        }
      }
    >
      Product 2
    </Text>
    <Text
      style={
        {
          "color": "#888",
          "fontSize": 14,
          "marginTop": 5,
        }
      }
    >
      Color: 
      Black
    </Text>
    <Text
      style={
        {
          "color": "#FF5722",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginTop": 5,
        }
      }
    >
      Price: $
      15
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FF5722",
          "borderRadius": 5,
          "flexDirection": "row",
          "marginTop": 10,
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 5,
        }
      }
    >
      <Text
        style={
          {
            "color": "#fff",
            "fontWeight": "bold",
            "marginLeft": 5,
          }
        }
      >
        Add to Cart
      </Text>
    </View>
  </View>
</View>
`;
